kernel/sched.c

```diff
@@ -767,14 +767,44 @@ bool z_thread_prio_set(struct k_thread *thread, int prio)
 		need_sched = z_is_thread_ready(thread);
 
 		if (need_sched) {
+#if defined(CONFIG_SMP)
+			int  old_prio;
+
+			old_prio = thread->base.prio;
+#endif /* CONFIG_SMP */
 			/* Don't requeue on SMP if it's the running thread */
 			if (!IS_ENABLED(CONFIG_SMP) || z_is_thread_queued(thread)) {
 				dequeue_thread(thread);
 				thread->base.prio = prio;
 				queue_thread(thread);
+#if defined(CONFIG_SMP)
```

Comment:
Can we clean up the #ifery here? This is sort of hard to read (i.e. an "#if CONFIG_SMP" inside an if() expression whose first term is "!IS_ENABLED(CONFIG_SMP)".  Pick one convention, don't mix them.  Maybe compute old_prio in such a way as to have it drop out in the optimizer, make flag_ipi() a noop when !SMP (is it not already?), etc...

Reply:
Actually, to be perfectly truthful: the more I look at this logic the less I like it.  Even ignoring the preprocessor-vs-C-logic issues, this is a lot of special-case code added just to handle the case of correctly optimizing IPI delivery for the case of a change in thread priority.   That's just not a performance case in all but the rarest apps; thread priorities are expected to be largely static while they're doing useful work.